library/src/main/java/org/apache/apex/malhar/lib/io/block/PartFileWriter.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.lib.io.block;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.validation.constraints.NotNull;
+
+import org.apache.commons.lang3.tuple.MutablePair;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.lib.io.block.AbstractBlockReader;
+import com.datatorrent.lib.io.block.BlockWriter;
+import com.datatorrent.lib.io.fs.AbstractFileSplitter;
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * Writes the blocks into the specified directory.
+ * If f1 is the file of size 10 MB and the block size is 1 MB then this operator writes the blocks into the
+ * specified directory as f1.part1, f1.part2 , ...., f1.part10. Here, size of each part is 1 MB.
+ */
+public class PartFileWriter extends BlockWriter implements Operator.IdleTimeHandler
+{
+  protected static String PARTSUFFIX = ".part";
+  @NotNull
+  private String outputDirectoryPath;
+  private Map<Long, MutablePair<Integer, String>> blockInfo = new HashMap<>();
+  private transient List<AbstractBlockReader.ReaderRecord<Slice>> waitingTuples;
+
+  public final transient DefaultInputPort<AbstractFileSplitter.FileMetadata> fileMetadataInput = new DefaultInputPort<AbstractFileSplitter.FileMetadata>()
+  {
+    @Override
+    public void process(AbstractFileSplitter.FileMetadata fileMetadata)
+    {
+      blockInfo.clear();
+      long[] blocks = fileMetadata.getBlockIds();
+      String relativePath = fileMetadata.getRelativePath();
+      for (int i = 0; i < blocks.length; i++) {
+        blockInfo.put(blocks[i], new MutablePair<>(i + 1, relativePath));
+      }
+    }
+  };
+
+  /**
+   * Initialize the file path
+   * @param context operator context
+   */
+  @Override
+  protected void initializeFilePath(Context.OperatorContext context)
+  {
+    filePath = outputDirectoryPath;
+    waitingTuples = new ArrayList<>();
+  }
+
+  @Override
+  protected void processTuple(AbstractBlockReader.ReaderRecord<Slice> tuple)
+  {
+    // Check whether the block id have index. If not, put it in waitingTuples.
+    if (blockInfo.get(tuple.getBlockId()) == null) {
+      waitingTuples.add(tuple);
+      return;
+    }
+    super.processTuple(tuple);
+  }
+
+  @Override
+  protected String getFileName(AbstractBlockReader.ReaderRecord<Slice> tuple)
+  {
+    MutablePair<Integer,String> blockId = blockInfo.get(tuple.getBlockId());
+    return blockId.getRight() + PARTSUFFIX + blockId.getLeft();
```

Comment:
blockId.getRight() returns the relative path of file and it consists of filename.